### PR TITLE
Added docs for Go API

### PR DESCRIPTION
**Pull request by merlimat**
Stacked on top of #239, #240

 * Added GoDoc for API, along with a few usage examples.
 * Added docs for keys sorting

### CODE REVIEW

File: oxia/options.go [hunk 1]
```diff
@@ -165,6 +151,7 @@ type putOptions struct {
 	ephemeral       bool
 }
 
+// PutOption represents an option for the [SyncClient.Put] operation
```
Comment (andrasbeni) [inline oxia/options.go:154]: And `AsyncClient.Put` also.

File: oxia/client.go [hunk 2]
```diff
@@ -21,73 +21,208 @@ import (
...
+// NotificationType represent the type of the notification event
 type NotificationType int
 
 const (
+	// KeyCreated A record that didn't exist was created
 	KeyCreated NotificationType = iota
+	// KeyModified An existing record was modified
 	KeyModified
+	// KeyDeleted A record was deleted
 	KeyDeleted
```
Comment (andrasbeni) [inline oxia/client.go:226]: A bit out of the scope of this change: based on the descriptions, it seems these should be named `RecordCreated`, `RecordModified`, and `RecordDeleted`.

File: cmd/client/notifications/cmd.go [hunk 3]
```diff
@@ -47,7 +47,7 @@ func exec(cmd *cobra.Command, args []string) error {
 		log.Info().
 			Stringer("type", notification.Type).
 			Str("key", notification.Key).
-			Int64("version", notification.Version).
+			Int64("version", notification.VersionId).
```
Comment (andrasbeni) [inline cmd/client/notifications/cmd.go]:
```suggestion
			Int64("version-id", notification.VersionId).
```

File: docs/oxia-key-sorting.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+
+# Oxia Keys Sorting
+
+Oxia uses a custom comparison operator for sorting the keys of the stored records. 
+
+In Oxia all the keys are independent of each other and there is no explicit tree-like structure
+with parent-children relations.
+
+Having said that, Oxia assign some special meaning to the `/` character in keys. This is done in
```
Comment (andrasbeni) [inline docs/oxia-key-sorting.md]:
```suggestion
Having said that, Oxia assigns a special meaning to the `/` character in keys. This is done in
```

File: docs/oxia-key-sorting.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+order to ensure efficient traversal of small portions of the key space.
+
+In the specific, Oxia ensure that keys with a common prefix and the same number of `/` segments
```
Comment (andrasbeni) [inline docs/oxia-key-sorting.md]:
```suggestion
Specifically, Oxia ensures that keys with a common prefix and the same number of `/` segments
```

File: oxia/client.go [hunk 6]
```diff
@@ -21,73 +21,208 @@ import (
...
 	ErrorUnexpectedVersionId = errors.New("unexpected version id")
-	ErrorUnknownStatus       = errors.New("unknown status")
+
+	// ErrorUnknownStatus Unknown error
+	ErrorUnknownStatus = errors.New("unknown status")
 )
 
+// AsyncClient Oxia client with methods suitable for asynchronous operations.
+//
+// This interface expose the same functionality as [SyncClient], though it returns
```
Comment (andrasbeni) [inline oxia/client.go]:
```suggestion
// This interface exposes the same functionality as [SyncClient], though it returns
```

File: oxia/client.go [hunk 7]
```diff
@@ -21,73 +21,208 @@ import (
...
+
+	// The version information
 	Version Version
-	Err     error
+
+	// The error if the `Get` operation failed
+	Err error
 }
 
+// ListResult structure is wrapping a list of keys, and an eventual error as
```
Comment (andrasbeni) [inline oxia/client.go]:
```suggestion
// ListResult structure is wrapping a list of keys, and a potential error as
```
to me eventual sounds like it's only a matter of time before you get an error. This does not suggest confidence.

File: oxia/client.go [hunk 8]
```diff
@@ -21,73 +21,208 @@ import (
...
+// results for a `List` operation in the [AsyncClient]
 type ListResult struct {
+	// The list of keys return by [List]
```
Comment (andrasbeni) [inline oxia/client.go]:
```suggestion
	// The list of keys returned by [List]
```

File: oxia/client.go [hunk 9]
```diff
@@ -21,73 +21,208 @@ import (
...
 	Keys []string
-	Err  error
+	// The eventual error in the [List] operation
+	Err error
 }
 
+// Notifications allows application to receive the feed of changes
```
Comment (andrasbeni) [inline oxia/client.go]:
```suggestion
// Notifications allow applications to receive the feed of changes
```

File: oxia/client.go [hunk 10]
```diff
@@ -21,73 +21,208 @@ import (
...
+// that are happening in the Oxia database
 type Notifications interface {
 	io.Closer
 
+	// Ch exposes the channel where all the notification events are published
 	Ch() <-chan *Notification
 }
 
+// NotificationType represent the type of the notification event
```
Comment (andrasbeni) [inline oxia/client.go]:
```suggestion
// NotificationType represents the type of the notification event
```

File: oxia/client.go (hunk 2 above)
Comment (merlimat) [reply oxia/client.go:226]: Yes, I didn't want to have many code changes in this PR

File: oxia/options.go (hunk 1 above)
Comment (merlimat) [reply oxia/options.go:154]: Yes, they're essentially the same operation, with two different flavors. I didn't want to be super-verbose there.

**Pull request merged**